onsenui/esm/elements/ons-popover.spec.js

```diff
@@ -271,6 +271,33 @@ describe('OnsPopoverElement', () => {
     });
   });
 
+  describe('\'target\' attribute', () => {
+    const targetElmId = 'test';
+    const popover = ons._util.createElement(`<ons-popover target="${targetElmId}">Test</ons-popover>`);
+    const target = ons._util.createElement(`<div id="${targetElmId}"></div>`);
+    document.body.appendChild(popover);
+    document.body.appendChild(target);
```

Comment:
It looks like this bit needs to be wrapped in `beforeEach` and then you need to use `afterEach` to remove the elements:

```javascript
let popover, target;
const targetElmId = 'test';

beforeEach(() => {
  popover = ons._util.createElement(`<ons-popover target="${targetElmId}">Test</ons-popover>`);
  target = ons._util.createElement(`<div id="${targetElmId}"></div>`);
  document.body.appendChild(popover);
  document.body.appendChild(target);
});

afterEach(() => {
  popover.remove();
  target.remove();
});

```

Reply:
@emccorson  Thank you for your comments.
I'm sorry, this might be a beginner question, but can I remove `contentReady(popover, done);` from  `beforeEach` and `popover = target = null;` from `afterEach` ?

```
  beforeEach(done => {
    popover = ons._util.createElement(`<ons-popover target="${targetElmId}">Test</ons-popover>`);
    target = ons._util.createElement(`<div id="${targetElmId}"></div>`);
    document.body.appendChild(popover);
    document.body.appendChild(target);

    contentReady(popover, done);
  });

  afterEach(() => {
    popover.remove();
    target.remove();
    popover = target = null;
  });
```

Reply:
@emccorson I applied your feedback. (Fixed test, rebase my branch to master, all commits into one commit. )
For now I keep the `contentReady(popover, done);` and `popover = target = null;` but please let me know what to do about them.

By the way, does the test of `master` fail? I hope this might not be due to my change.

```
Chrome Headless 105.0.5195.125 (Mac OS 10.15.7) OnsNavigatorElement #pushPage() push none exist page FAILED
        Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
Chrome Headless 105.0.5195.125 (Mac OS 10.15.7): Executed 577 of 982 (1 FAILED) (5.797 secs / 5.33 secs)
TOTAL: 1 FAILED, 576 SUCCESS
[15:29:52] Karma server has exited with 1
[15:29:52] Failed to pass some unit tests. (Otherwise, the unit testing itself is broken)
```

Reply:
`done` is used with promises to signal that the setup is done so we can start running the tests. `contentReady` is similar to `setImmediate` i.e. it is used to wait for the things to be set up. `contentReady` has a queue of actions which it processes in order when ready, so creating the ons-popover will add some set up code to the queue, and `contentReady(popover, done)` is processed after this. Basically it is a way of signalling that all the necessary set up has been done.

However, there are big problems with `contentReady` (see v3 doc for an explanation). It should probably be replaced, though both me and Matagawa-san tried to replace it separately, and it was very hard both times.

For the tests, you can probably use `setImmediate` instead of `contentReady` (`contentReady` eventually uses `setImmediate` as a [fallback](https://github.com/OnsenUI/OnsenUI/blob/master/onsenui/esm/ons/content-ready.js#L64) anyway).

---

The tests failed because the Safari service failed to start on the CI. This happens sometimes, but re-running the tests usually fixes it. Passing now.